src/ldap_auth.rs

```diff
@@ -1,21 +1,46 @@
-use std::io;
+use std::ptr;
 
-use ldap3::{LdapConn, Scope, SearchEntry, ldap_escape, parse_filter};
+use openldap::{self, LDAPResponse, RustLDAP};
+
+use regex::Regex;
 
 use config::LdapConfig;
+use errors::*;
+
+pub struct LDAPEntry {
+    pub dn: String,
+}
+
+fn new_ldap(url: &str) -> Result<RustLDAP> {
+    let ldap = RustLDAP::new(url)?;
+
+    ldap.set_option(
+        openldap::codes::options::LDAP_OPT_PROTOCOL_VERSION,
+        &openldap::codes::versions::LDAP_VERSION3,
+    );
 
-pub fn auth(user: &str, pass: &str, config: &LdapConfig) -> io::Result<bool> {
+    ldap.set_option(
+        openldap::codes::options::LDAP_OPT_X_TLS_REQUIRE_CERT,
+        &openldap::codes::options::LDAP_OPT_X_TLS_DEMAND,
+    );
+
+    Ok(ldap)
+}
+
+pub fn auth(user: &str, pass: &str, config: &LdapConfig) -> Result<bool> {
     if user.is_empty() {
         info!("Cannot authenticate without username");
         return Ok(false);
     }
 
-    let user_safe = ldap_escape(user);
-    let user_filters = config
-        .userid_attributes
-        .iter()
-        .map(|a| format!("({}={})", a, user_safe.as_ref()))
-        .collect::<Vec<_>>();
+    // in the absence of `ldap_escape` from ldap3, just whitelist acceptable characters
+    let re = Regex::new(r"([^A-Za-z0-9\.\-_@])").unwrap();
```

Comment:
I couldn't find a list of characters in the RFC. Is there a reference you listed them from?

Reply:
No, wasn't referencing an RFC. Was just sticking to alphanumeric, dots, dashes and `@` (for `userPrincipalName`'s)